disk-utils/cfdisk.c

```diff
@@ -2752,19 +2761,34 @@ int main(int argc, char *argv[])
 	} else
 		diskpath = argv[optind];
 
-	rc = fdisk_assign_device(cf->cxt, diskpath, 0);
-	if (rc == -EACCES)
-		rc = fdisk_assign_device(cf->cxt, diskpath, 1);
+        uid = getuid();
+        if (!read_only){
+	        rc = fdisk_assign_device(cf->cxt, diskpath, 0);
+
+                /*Forced try to open in read-only mode */
+	        if (rc == -EACCES){
+                        struct stat disk_perms;
+                        if (uid){
+                                if (!stat(diskpath, &disk_perms)){
+                                        if(!(disk_perms.st_mode & S_IROTH))
+                                                err(EXIT_FAILURE, _("reading of %s"), diskpath);
+                                }
+                                rc = fdisk_assign_device(cf->cxt, diskpath, 1);
+                        }
+                        else
+                                err(EXIT_FAILURE, _("writing to %s denied to root"), diskpath);
+                }
+                
+        } else {
+                rc = fdisk_assign_device(cf->cxt, diskpath, 1);
+        }
 	if (rc != 0)
```

Comment:
This is overkill and I have doubts it's correct :-)

You can get EACCES for more reasons (like SELinux, insufficient capabilities, etc.).  It seems fragile to implement any policy to cfdisk based on st_mode. It's usually better to try open() and really on kernel reply than predict anything.

IMHO all you need is:
```C
rc = fdisk_assign_device(cf->cxt, diskpath, read_only);
if (rc == -EACCES && read_only == 0)
  rc = fdisk_assign_device(cf->cxt, diskpath, 1);
if (rc != 0)
  err(EXIT_FAILURE, _("cannot open %s"), diskpath);
```